tsec-http4s/src/main/scala/tsec/authentication/credentials/CredentialStore.scala

```diff
@@ -11,24 +12,76 @@ import tsec.passwordhashers.jca._
   */
 trait CredentialStore[F[_], C, P] {
 
-  def putCredentials(credentials: C): F[Unit]
+  /** Add credentials to some backing store.
+    * `putCredentials` takes a function, most likely
+    * partially applied on some userId, that takes a `P`
+    * and adds some user record onto a credentials store.
+    *
+    * @param credentials
+    * @param put
+    * @return
+    */
+  def putCredentials(credentials: C, put: P => F[Unit]): F[Unit]
 
-  def updateCredentials(credentials: C): F[Unit]
-
-  def removeCredentials(credentials: C): F[Unit]
+  /** Alias for `putCredentials`
+    * for when anyone may want to be verbose about
+    * put vs update
+    */
+  def updateCredentials(credentials: C, update: P => F[Unit]): F[Unit] =
+    putCredentials(credentials, update)
 
+  @deprecated("Use isAuthenticated", "0.0.1-M10")
   def authenticate(credentials: C): F[Boolean]
+
+  def isAuthenticated(credentials: C): F[Boolean]
 }
 
-abstract class PasswordStore[F[_]: Sync, Id, P](implicit h: PasswordHasher[F, P])
-    extends CredentialStore[F, RawCredentials[Id], P] {
+abstract class PasswordStore[F[_], Id, P](implicit P: PasswordHasher[F, P], F: Sync[F])
+    extends CredentialStore[F, RawCredentials[Id], PasswordHash[P]] {
 
-  def retrievePass(id: Id): F[PasswordHash[P]]
+  def retrievePass(id: Id): OptionT[F, PasswordHash[P]]
 
+  @deprecated("Use isAuthenticated", "0.0.1-M10")
   def authenticate(credentials: RawCredentials[Id]): F[Boolean] =
+    isAuthenticated(credentials)
+
+  def putCredentials(credentials: RawCredentials[Id], put: PasswordHash[P] => F[Unit]): F[Unit] =
+    for {
+      hash <- P.hashpw(credentials.rawPassword)
+      _    <- put(hash)
+    } yield ()
+
+  def putCredentials(raw: Array[Byte], put: PasswordHash[P] => F[Unit]): F[Unit] =
+    for {
+      hash <- P.hashpw(raw)
+      _    <- put(hash)
+    } yield ()
+
+  def putCredentials(raw: Array[Char], put: PasswordHash[P] => F[Unit]): F[Unit] =
+    for {
+      hash <- P.hashpw(raw)
+      _    <- put(hash)
+    } yield ()
+
+  def isAuthenticated(credentials: RawCredentials[Id]): F[Boolean] =
+    for {
+      pass <- retrievePass(credentials.identity)
+        .getOrElseF(F.raiseError(new NoSuchElementException("No such user")))
```

Comment:
Why not use https://github.com/jmcardon/tsec/blob/master/common/src/main/scala/tsec/common/package.scala#L15